Issue#56: Implement exception handling

isvyry opened the pull request:
## ZenHab

* [Main ZenHab ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/56)


## Code reviewers

- [ ] @alexorlenko
- [ ] @Arthurpanda  
- [ ] @mykytam 
- [ ] @omelchenkotimur  
- [ ] @reeztl 

### Second Level Review

- [ ] @DzigMS 
- [ ] @kh0ma 

## Summary of issue

Implement initial exception handling in our application

## Summary of change

An error message is sent with exception details when it is caught

## Testing approach

---

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

isvyry committed:
Issue#56: Implement exception handling
Refactor JerseyConfig so it is able to scan exception mappers and register them
Add AlreadyExistException and NotFoundException mappers

isvyry committed:
Issue#56: Implement exception handling
Rename Error component to ApiError in open api specification file
Delete description field from ApiError
Set responseCode field type to integer in ApiError
Refactor mapper so it uses generated ApiError

mykytam committed:
Issue#56: Implement exception handing
Add documentation to classes

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/AlreadyExistExceptionMapper.java:
```diff
@@ -0,0 +1,28 @@
+ * @author Ihor Svyrydenko
+ */
+@Provider
+public class AlreadyExistExceptionMapper implements ExceptionMapper<AlreadyExistException> {
```
Please introduce abstract BaseApiException. Extend `AlreadyExistException` and `NotFoundException` from it. Also pay attention to the naming of the exception, in this case, would be better to name it `NotFoundApiException`.

From the mapper's perspective please handle the mapping to the response in the abstract BaseApiExceptionMapper. Move the changeable data to the abstract methods.

mykytam replied:
Ok, noted. Changes will be commited.

mykytam committed:
Issue#56: Implement exception handling
Add BaseApiExceptionMapper to handle the mapping to the response
Rename exceptions
Rewrite mappers due to abstract class

mykytam committed:
Merge remote-tracking branch 'origin/dev' into Issue#56-Implement-exception-handling

# Conflicts:
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/AlreadyExistException.java
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/HomeException.java
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/NotFoundException.java

Dandria1802 committed:
Refactor exception mappers

Add GlobalExceptionMapper
Refactor:added access modifiers

Dandria1802 committed:
Merge remote-tracking branch 'origin/dev' into Issue#56-Implement-exception-handling

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/GlobalExceptionMapper.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public Response toResponse(Exception exception) {
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
```
Is this mapping following our contract in open API definition?

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/exception/BaseApiException.java:
```diff
@@ -0,0 +1,8 @@
+package com.softserveinc.ita.homeproject.homeservice.exception;
+
+public abstract class BaseApiException extends RuntimeException {
```
The service layer should know nothing about presentation so we cannot use anything from API here. From the package I see that this exception belongs to service layer, please remove API mentioning from the naming:
```
BaseHomeException
  NotFoundHomeException
  AlreadyExsistHomeException
```

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/BaseApiExceptionMapper.java:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @author Mykyta Morar
+ */
+public abstract class BaseApiExceptionMapper<T extends BaseApiException>  implements ExceptionMapper<T> {
```
Please consider the following structure:
```java
public abstract class BaseExceptionMapper<T extends Throwable> implements ExceptionMapper<T> {

    @Override
    public Response toResponse(T exception) {
    	Response.Status status = getStatus();
        return Response.status(status)
                .entity(new ApiError()
                        .responseCode(status.getStatusCode())
                        .errorMessage(extractMessage(exception))
                .build();
    }

    protected abstract Response.Status getStatus();
        
    protected abstract String extractMessage(T exception);

}

public class GeneralExceptionMapper extends BaseExceptionMapper<Trowable> {

    protected Response.Status getStatus() {
    	return INTERNAL_SERVER_ERROR;
    }
        
    protected String extractMessage(Trowable exception) {
    	return "Unknown error. Please contact support.";
    }

}

public abstract class BaseHomeExceptionMapper<T extends BaseHomeException> extends BaseExceptionMapper<T> {
        
    protected String extractMessage(T exception) {
    	return exception.getMessage();
    }

}


public class NotFoundHomeExceptionMapper extends BaseHomeExceptionMapper<NotFoundHomeException> {

    protected Response.Status getStatus() {
    	return ...; //404
    }

}


public class AlreadExistHomeExceptionMapper extends BaseHomeExceptionMapper<AlreadExistHomeException> {

    protected Response.Status getStatus() {
    	return ...; //400
    }

}
```

Dandria1802 committed:
Refactor exception mappers

Add ВaseHomeExceptionMapper
Refactor: renamed exception classes

Dandria1802 committed:
Merge remote-tracking branch 'origin/dev' into Issue#56-Implement-exception-handling

# Conflicts:
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/NewsServiceImpl.java
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java

Dandria1802 committed:
Issue#56 Fixed merge conflicts

Dandria1802 committed:
Issue#56 Add new mapper NotFoundException

Dandria1802 committed:
Merge remote-tracking branch 'origin/dev' into Issue#56-Implement-exception-handling

# Conflicts:
#	home-open-api/src/main/resources/yaml/openapi.yaml
#	home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/NewsServiceImpl.java

Dandria1802 committed:
Issue#56 Fixed merge conflicts
